### PR TITLE
chapter 1 and 2 solutions

### PR DESCRIPTION
**Pull request by minimario**
### Solutions for Chapter 1

Hi @vrom911 @chshersh, thanks for putting this together! I'd love to you to take a look at my solutions and suggest some improvements! I've been playing around with Haskell for a few months, but I'm not too familiar with best practices/coding idiomatically :)

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -653,7 +667,11 @@ You need to use recursion in this task. Feel free to return to it later, if you
 aren't ready for this boss yet!
 -}
 
-firstDigit n = error "firstDigit: Not implemented!"
+firstDigit :: Int -> Int
+firstDigit n
+  | n < 0 = firstDigit (-n)
+  | n < 10 = n
+  | otherwise = firstDigit $ div n 10
```
Comment (vrom911) [inline src/Chapter1.hs:674]: Neat advanced task solution 👏🏼

File: src/Chapter2.hs [hunk 2]
```diff
@@ -336,7 +347,9 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList l r xs =
+  if l<0 || r<0 || l>r then []
```
Comment (vrom911) [inline src/Chapter2.hs:351]: In Haskell, functions and operators are always separated from arguments with spaces 👾 

```suggestion
  if l < 0 || r < 0 || l > r then []
```

File: src/Chapter2.hs [hunk 3]
```diff
@@ -336,7 +347,9 @@ from it!
...
+  else take (r-l+1) $ drop l xs
```
Comment (vrom911) [inline src/Chapter2.hs:352]: And here as well 🙂 
```suggestion
  else take (r - l + 1) $ drop l xs
```

File: src/Chapter2.hs [hunk 4]
```diff
@@ -348,8 +361,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+
+firstHalf :: [a] -> [a]
+firstHalf l =
+  take len l where len = div (length l) 2
```
Comment (vrom911) [inline src/Chapter2.hs:367]: Nice usage of `where` construction 👌🏼

File: src/Chapter2.hs [hunk 5]
```diff
@@ -501,8 +516,9 @@ True
 >>> isThird42 [42, 42, 0, 42]
 False
 -}
-isThird42 = error "isThird42: Not implemented!"
-
+isThird42 :: [Int] -> Bool
+isThird42 (_:_:a:_) = (a == 42)
```
Comment (vrom911) [inline src/Chapter2.hs:520]: You don't need to compare `z` with 42 separately, you can pattern match on `42` directly in the pattern.

```suggestion
isThird42 (_:_:42:_) = True
```

File: src/Chapter2.hs [hunk 6]
```diff
@@ -621,7 +637,11 @@ Write a function that takes elements of a list only in even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+
+takeEven :: [Int] -> [Int]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven (x:_:xs) = x:takeEven xs
```
Comment (vrom911) [inline src/Chapter2.hs:644]: Great pattern matching skills 💪🏼

File: src/Chapter2.hs [hunk 7]
```diff
@@ -728,7 +748,8 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = 
+  concat $ map (\x->replicate x x) l
```
Comment (vrom911) [inline src/Chapter2.hs:752]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 8]
```diff
@@ -741,7 +762,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains e xs = filter (\x -> elem e x) xs
```
Comment (vrom911) [inline src/Chapter2.hs:766]: Now, after you've mastered eta-reduction, you can apply such a technique to this function and the lambda function inside as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 9]
```diff
@@ -781,13 +803,13 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
-divideTenBy x = div 10 x
+divideTenBy = div 10
 
 -- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:812]: Eta-reduction award :trophy:

File: src/Chapter2.hs [hunk 10]
```diff
@@ -858,7 +883,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = (rewind xs) ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs:888]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Fantastic work! 👏🏼 
Congrats with finishing two Chapters 🎉

**Comment by chshersh** [comment]
@minimario I'm going to close this PR, as we can't merge it to our repo 🙂

**Pull request closed without merging**